drivers/framework.zig

```diff
@@ -42,6 +43,128 @@ pub const wireless = struct {
+        }
+
+        pub fn to_us(abs: Absolute) u64 {
+            return @intFromEnum(abs);
+        }
+
+        pub fn is_reached_by(deadline: Absolute, point: Absolute) bool {
+            return deadline.to_us() <= point.to_us();
+        }
+
+        pub fn diff(future: Absolute, past: Absolute) Duration {
+            return Duration.from_us(future.to_us() - past.to_us());
+        }
+
+        pub fn add_duration(abs: Absolute, dur: Duration) Absolute {
+            return Absolute.from_us(abs.to_us() + dur.to_us());
+        }
+    };
+
+    ///
+    /// A duration with microsecond precision.
+    ///
+    /// NOTE: Using an `enum` type here prevents type confusion with other
+    ///       related or unrelated integer-like types.
+    ///
+    pub const Duration = enum(u64) {
+        _,
+
+        pub fn from_us(us: u64) Duration {
+            return @as(Duration, @enumFromInt(us));
+        }
+
+        pub fn from_ms(ms: u64) Duration {
+            return from_us(1000 * ms);
+        }
+
+        pub fn to_us(duration: Duration) u64 {
+            return @intFromEnum(duration);
+        }
+
+        pub fn less_than(self: Duration, other: Duration) bool {
+            return self.to_us() < other.to_us();
+        }
+
+        pub fn minus(self: Duration, other: Duration) Duration {
+            return from_us(self.to_us() - other.to_us());
+        }
+
+        pub fn plus(self: Duration, other: Duration) Duration {
+            return from_us(self.to_us() + other.to_us());
+        }
+    };
+
+    ///
+    /// The deadline construct is a construct to create optional timeouts.
+    ///
+    /// NOTE: Deadlines use maximum possible `Absolute` time for
+    ///       marking the deadline as "unreachable", as this would mean the device
+    ///       would ever reach an uptime of over 500.000 years.
+    ///
+    pub const Deadline = struct {
+        const disabled_sentinel = Absolute.from_us(std.math.maxInt(u64));
+
+        timeout: Absolute,
+
+        /// Create a new deadline with an absolute end point.
+        ///
+        /// NOTE: `abs` must not point to the absolute maximum time stamp, as this is
+        ///       used as a sentinel for "unset" deadlines.
+        pub fn init_absolute(abs: ?Absolute) Deadline {
+            if (abs) |a|
+                std.debug.assert(a != disabled_sentinel);
+            return .{ .timeout = abs orelse disabled_sentinel };
+        }
+
+        /// Create a new deadline with a certain duration from provided time.
+        pub fn init_relative(since: Absolute, dur: ?Duration) Deadline {
+            return init_absolute(if (dur) |d|
+                make_timeout(since, d)
+            else
+                null);
+        }
+
+        /// Returns `true` if the deadline can be reached.
+        pub fn can_be_reached(deadline: Deadline) bool {
+            return (deadline.timeout != disabled_sentinel);
+        }
+
+        /// Since this takes 'now', rename to is_reached_by?
```

Comment:
probably should do this, I do it above in `Absolute`.

Would be nice if the calls that use this didn't have to call `time.get_time_since_boot()` (or equivalent), but I am can't think of a nice way to do that. Maybe the rp2xxx hal can make a wrapping type that adds methods?